### ISSUE TITLE
[Optimization] meson "unity" builds and `-fwhole-program`

### ISSUE THREAD

**Issue by pchome** [InBetweenNames/gentooLTO#151]
There is still few packages in gentoo portage tree that using meson build system, but anyway ...

### meson unity builds
[unity-builds](http://mesonbuild.com/Unity-builds.html#unity-builds)
> Normally we would run the compiler three times, once for each file. In a unity build we instead compile all these sources in a single unit.

> The exact speedup depends on the code base, of course, but it is not uncommon to obtain more than 50% speedup in compile times. This happens even though the Unity build uses only one CPU whereas individual compiles can be run in parallel. Unity builds can also lead to faster code, because the compiler can do more aggressive optimizations (e.g. inlining).

Maybe such feature is not important for LTO, but it opens room for `-fwhole-program` optimization for packages currently lacks LTO support. Moreover `-fwhole-program` do not required for LTO, but still can be used (AFAIK, man gcc).

<hr/>

### meson wrap
[Wrap dependency system](http://mesonbuild.com/Wrap-dependency-system-manual.html)

For packages w/o meson support.
TL;DR: you can create `meson.build` files for any project, redistribute them separately (e.g. https://wrapdb.mesonbuild.com/), and then build project using `meson` and it's features.

<hr/>

### Example
[DXVK](https://github.com/doitsujin/dxvk): project using meson build system, and [here](https://github.com/pchome/dxvk-gentoo-overlay/blob/master/games-util/dxvk/dxvk-9999-r1.ebuild) I have an ebuild for it (winelib build), which
* using `--unity=on` meson feature by default
* have an "hack" for build system (see [patches](https://github.com/pchome/dxvk-gentoo-overlay/tree/master/games-util/dxvk/files) and `USE=experimental`) to use `-fwhole-program` optimization option on modified "unity files".

lib          | system       | test          | default (meson --buildtype "release")
------------ | ------------ | ------------- | -------------
dxgi.dll.so | 931592 | 632680 | 620296
d3d11.dll.so | 1728584 | 1118024 | 1076968
d3d10core.dll.so | 183944 | 117728 | 113600
d3d10.dll.so | 188648 | 139536 | 135432
d3d10_1.dll.so | 188648 | 123040 | 123024

`system` mean `-O3`+graphite (system-wide installed package)
`test` mean `system`+`-fwhole-program`
`default` mean `-O3`+`-fwhole-program`

At least this feature noticeable affects binary sizes. DXVK is not fully suitable for CPU optimization tests.

_We can test something like this on no-LTO packages.
In addition to #150_

**Comment by InBetweenNames** [comment]
Indeed, a unity build is really a poor man's LTO!  All source files are concatenated into a single TU, which is then compiled monolithically.  The results should be exactly the same as using LTO, barring some restrictions. In particular, it has the disadvantages of not being able to inline across static library boundaries or other programming languages.  However, I think it is worth looking at for meson, since it is a particularly problematic build system when it comes to `*FLAGS`, as you have noted!  Will investigate this more in the coming weeks.

**Comment by cryptopsy0** [comment]
does this relate to how qtwebengine USE=jumbo-build? I have read somewhere along the lines that with -j12 it uses 12gb ram + 8gb swap. this is too much for my for my system using. i get nearly 7.5gb memory usage compiling qtwebengine with -j4 and it seems to grow as they release new versions. i compile in a tmpfs /tmp to speed building up.

**Comment by InBetweenNames** [comment]
@cryptopsy0 , `jumbo-build` is indeed a unity build.  And as @pchome said, the more threads you throw at it, the more memory you'll need.  I'm not sure how unity builds interact with LTO in general, but I would expect the benefits of LTO to be diminished in jumbo builds as there are fewer TUs.  That said, both should help the compiler perform interprocedural optimizations across boundaries.  With the proposed approach, we would prefer, in order:

* LTO
* Unity build with -fwhole-program
* -ffat-lto-objects
* no LTO
